Add schemas for plans

thrix opened the pull request:
Todo
- [x] context
- [x] discover
- [x] prepare
- [x] provision
- [x] execute
- [x] finish
- [x] report

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

thrix commented:
@happz @lukaszachy @psss this now passes for me and validates all available plans.

One thing that is fairly unclear is the role of the core attributes, I would expect they can be part of any fmf node, i.e. a plan, test or story. But according to @lukaszachy they can be in any leaf of the node data??

thrix commented on tests/unit/test_schemas.py:
```diff
@@ -27,16 +27,36 @@ def stories_schema():
         schema_file = os.path.join(SCHEMADIR, f'{schema_name}.yaml')
-        schema = YAML(typ="safe").load(
-            open(schema_file, encoding="utf-8"))
+        schema = YAML(
```
formatting

thrix commented on tmt/schemas/common.yaml:
```diff
@@ -148,11 +214,41 @@ definitions:
   # a url, path or fmf_id
   url_path_or_fmf_id:
     oneOf:
-      - type: string
```
hmm, what the hell is this change :)

thrix replied:
nm, it is correct, url become a common definition.

guy9050 commented on tests/unit/test_schemas.py:
```diff
@@ -27,13 +27,30 @@ def stories_schema():
     return YAML(typ="safe").load(open(schema_file, encoding="utf-8"))
 
 
+@pytest.fixture
```
Would it make sense to set a scope to this fixture?

happz replied:
Probably not. Fixture scope takes care of clearing fixture when pytest leaves the given scope (module, package, etc.), and this particular fixture is not harmless, does not spoil other tests or fixtures, and it's not in conflict with any other fixture.

thrix replied:
thanks Milos.

thrix commented:
@psss @lukaszachy @happz this one needs finally review to unblock #1178

I had one discussion topic, but I moved it to #1182 discussion topics, and I guess we will discuss it briefly on the hacking session.

lukaszachy commented:
@thrix Please add 'id' to the schema (it was added to the core attributes just recently)

psss commented on tmt/schemas/provision/hardware.yml:
```diff
@@ -0,0 +1,220 @@
+    - "$ref": "#/definitions/hw_virtualization"
+    - "$ref": "#/definitions/hw_block"
+    - "$ref": "#/definitions/hw_and"
+    - "$ref": "#/definitions/hw_or"
```
The `memory` key is missing here. Tested with a simple plan example:

```yaml
provision:
    how: artemis
    hardware:
        memory: 8GB
```

Aren't the key names missing here? Shouldn't the definition here be similar as in `hw_block` above?

happz replied:
I'm not sure, it looks correct, but I am too seeing validation errors with your example. I don't know why, yet, but the schema should work, it's been used by Artemis for months. Maybe there's a typo or a copy/paste error.

happz replied:
With this patch to hardware schema, things start to work:

```patch
diff --git a/tmt/schemas/provision/artemis.yaml b/tmt/schemas/provision/artemis.yaml
index b7c8397..6fabcbb 100644
--- a/tmt/schemas/provision/artemis.yaml
+++ b/tmt/schemas/provision/artemis.yaml
@@ -51,7 +51,7 @@ properties:
     type: string
 
   hardware:
-    $ref: "/schemas/provision/hardware#"
+    $ref: "/schemas/provision/hardware#/definitions/hardware"
 
   pool:
     type: string
diff --git a/tmt/schemas/provision/hardware.yaml b/tmt/schemas/provision/hardware.yaml
index be79a7a..b4a58f4 100644
--- a/tmt/schemas/provision/hardware.yaml
+++ b/tmt/schemas/provision/hardware.yaml
@@ -203,18 +203,13 @@ definitions:
     required:
       - "or"
 
-type: object
-additionalProperties: false
-
-properties:
-  anyOf:
-    - "$ref": "#/definitions/hw_boot"
-    - "$ref": "#/definitions/hw_cpu"
-    - "$ref": "#/definitions/hw_disks"
-    - "$ref": "#/definitions/hw_networks"
-    - "$ref": "#/definitions/hw_virtualization"
-    - "$ref": "#/definitions/hw_block"
-    - "$ref": "#/definitions/hw_and"
-    - "$ref": "#/definitions/hw_or"
-
-  required: []
+  hardware:
+    anyOf:
+      - "$ref": "#/definitions/hw_boot"
+      - "$ref": "#/definitions/hw_cpu"
+      - "$ref": "#/definitions/hw_disks"
+      - "$ref": "#/definitions/hw_networks"
+      - "$ref": "#/definitions/hw_virtualization"
+      - "$ref": "#/definitions/hw_block"
+      - "$ref": "#/definitions/hw_and"
+      - "$ref": "#/definitions/hw_or"
```

1. add a definition of the whole `hardware` key as a definition, not an object described by the schema in `hardware.yaml`
2. this definition is not yet another object but uses `anyOf` directly, it "becomes" one of the listed schemas as needed.

The schema in this PR was probably introducing one extra level of objects by putting this `anyOf` under `properties` in `hardware.yaml`, and this was the issue.

thrix replied:
@happz thanks for the investigation and solution, I applied it.

psss commented on tmt/schemas/provision/virtual.yaml:
```diff
@@ -0,0 +1,35 @@
+
+  user:
+    type: string
+
```
Missing keys `disk`, `connection` and `arch`.

thrix replied:
@psss added, please check

psss commented on tmt/schemas/report/html.yaml:
```diff
@@ -0,0 +1,23 @@
+  how:
+    type: string
+    enum:
+      - html
```
Missing key `open`.

thrix replied:
added as boolean

psss reviewed (dismissed):
Thanks much for preparing this. This must have been a lot of work! Looks good in general. Added a bunch of comments and questions.

psss commented on tests/unit/test_schemas.py:
```diff
@@ -26,13 +26,30 @@ def stories_schema():
+            'execute/tmt', 'execute/upgrade',
+            'finish/ansible', 'finish/shell',
+            'prepare/ansible', 'prepare/install', 'prepare/shell',
+            'provision/artemis', 'provision/connect', 'provision/container',
```
Missing `provision/hardware`. The file should be renamed from `hardware.yml` to `hardware.yaml`.

happz replied:
(FYI, https://github.com/teemtee/tmt/pull/1266 should take care of the "we forgot to list a file" issue)

thrix replied:
@psss thanks, added and renamed

thrix commented:
> @thrix Please add 'id' to the schema (it was added to the core attributes just recently)

@lukaszachy thanks, added

thrix commented on tmt/schemas/common.yaml:
```diff
@@ -8,6 +8,54 @@ $id: /schemas/common
+        $ref: "/schemas/core#/definitions/order"
+
+      playbook:
+        $ref: "/schemas/common#/definitions/array_of_strings"
```
thanks!

thrix commented on tmt/schemas/common.yaml:
```diff
@@ -148,11 +214,41 @@ definitions:
         items:
           type: string
 
+  # common definition of an ansible execution method
```
thanks, fixed.

thrix commented on tmt/schemas/discover/shell.yaml:
```diff
@@ -0,0 +1,50 @@
+          type: string
+
+        duration:
+          $ref: "/schemas/common#/definitions/duration"
```
I will try, the problem is the `name` which only appears here.

thrix replied:
I linked all attributes, there is no better way

thrix commented on tmt/schemas/plan.yaml:
```diff
@@ -0,0 +1,150 @@
+      - $ref: "/schemas/execute/upgrade#"
+      - type: array
+        items:
+          $ref: "/schemas/execute/tmt#"
```
ack. added

thrix commented on tmt/schemas/plan.yaml:
```diff
@@ -0,0 +1,150 @@
+  finish:
+    oneOf:
+      - $ref: "/schemas/finish/ansible#"
+      - $ref: "/schemas/finish/shell#"
```
thanks, added

thrix commented on tmt/schemas/plan.yaml:
```diff
@@ -0,0 +1,150 @@
+    $ref: "/schemas/core#/definitions/order"
+
+  # https://tmt.readthedocs.io/en/stable/spec/plans.html#finish
+  # TODO: can be a list
```
indeed, removed

thrix commented on tmt/schemas/plan.yaml:
```diff
@@ -0,0 +1,150 @@
+  order:
+    $ref: "/schemas/core#/definitions/order"
+
+  # https://tmt.readthedocs.io/en/stable/spec/plans.html#finish
```
thanks, fixed

thrix commented on tmt/schemas/plan.yaml:
```diff
@@ -0,0 +1,150 @@
+  report:
+    oneOf:
+      - $ref: "/schemas/report/display#"
+      - $ref: "/schemas/report/html#"
```
thanks, added

thrix commented on tmt/schemas/provision/connect.yaml:
```diff
@@ -0,0 +1,44 @@
+
+  port:
+    type: integer
+    multipleOf: 1
```
removed all occurances

thrix commented on tmt/schemas/plan.yaml:
```diff
@@ -0,0 +1,150 @@
+            - $ref: "/schemas/prepare/shell#"
+
+  # https://tmt.readthedocs.io/en/stable/spec/plans.html#provision
+  # TODO: can be a list
```
added array variant, removed TODO

thrix commented on tmt/schemas/provision/artemis.yaml:
```diff
@@ -0,0 +1,82 @@
+
+  api-retries:
+    type: integer
+    multipleOf: 1
```
Removed, all occurrences, not sure why I thought it makes any sense

thrix commented on tmt/schemas/provision/container.yaml:
```diff
@@ -0,0 +1,29 @@
+    type: string
+
+  image:
+    type: string
```
Added

thrix commented on tmt/schemas/provision/hardware.yml:
```diff
@@ -0,0 +1,220 @@
+definitions:
+
+  # HW requirements: `boot` block
+  hw_boot:
```
removed prefix, I agree it looks better without it. It was kinda copy paste error, I stole this from Artemis.

thrix commented on tmt/schemas/provision/hardware.yml:
```diff
@@ -0,0 +1,220 @@
+
+    properties:
+      type:
+        type: string
```
Added them, I would add them now. They are documented.

lukaszachy commented on tmt/schemas/plan.yaml:
```diff
@@ -0,0 +1,150 @@
+        $ref: "/schemas/common#/definitions/arch"
+
+      component:
+        type: string
```
Yes, fmf is treating all dimensions same way so the value can be one or more strings.

thrix replied:
yes, @lukaszachy confirmed on IRC ...

thrix replied:
it should be fixed

thrix commented:
> Thanks much for preparing this. This must have been a lot of work! Looks good in general. Added a bunch of comments and questions.

Thanks for the great review and good eyes, I knew I had to miss some things. Should have looked more on module code, I actually used the documentation, which is not actually mentioning some attributes. I believe it is ready for the next round.

CC @lukaszachy @happz

thrix commented on tmt/schemas/common.yaml:
```diff
@@ -23,6 +71,17 @@ definitions:
+  # https://tmt.readthedocs.io/en/stable/spec/plans.html#shell
+  duration:
+    type: string
+    pattern: "^[0-9]+[smhd]$"
```
made optional

thrix commented on tmt/schemas/discover/fmf.yaml:
```diff
@@ -0,0 +1,57 @@
+    $ref: "/schemas/common#/definitions/one_or_more_strings"
+
+  link:
+    $ref: "/schemas/common#/definitions/array_of_strings"
```
Filed an issue:| https://github.com/teemtee/tmt/issues/1374

thrix commented on tmt/schemas/discover/fmf.yaml:
```diff
@@ -0,0 +1,57 @@
+
+  modified-ref:
+    type: string
+
```
Thanks! added all

thrix commented on tmt/schemas/discover/shell.yaml:
```diff
@@ -0,0 +1,123 @@
+      patternProperties:
+        ^extra-:
+          type: string
+
```
thanks, added!

thrix commented on tmt/schemas/discover/shell.yaml:
```diff
@@ -0,0 +1,123 @@
+      additionalProperties: false
+
+      properties:
+
```
it is not possible, we had the same problem including core attributes

thrix commented on tmt/schemas/execute/upgrade.yaml:
```diff
@@ -0,0 +1,33 @@
+    enum:
+      - upgrade
+
+  filter:
```
Added all

psss commented on tmt/schemas/plan.yaml:
```diff
@@ -0,0 +1,199 @@
+            - $ref: "/schemas/provision/local#"
+            - $ref: "/schemas/provision/virtual#"
+
+  # https://tmt.readthedocs.io/en/stable/spec/plans.html#provision
```
```suggestion
  # https://tmt.readthedocs.io/en/stable/spec/plans.html#report
```

thrix replied:
fixed

psss commented on tmt/schemas/plan.yaml:
```diff
@@ -0,0 +1,199 @@
+  report:
+    oneOf:
+      - $ref: "/schemas/report/display#"
+      - $ref: "/schemas/report/html#"
```
We're missing the `junit` report plugin.

thrix replied:
added junit plugin

psss reviewed (changes requested):
Thanks for addressing all comments. Added two more.

thrix commented:
@lukaszachy @happz @psss addressed all concerns! Thanks for all the comments, feel free to take a look again, happy this finally gets in and unblocks base.py typing

psss reviewed (approved):
Thanks for addressing all comments! Looks good.

thrix commented:
oh my, forgot to squash
